### PR TITLE
Paginate service connect telemetry data

### PR DESCRIPTION
**Pull request by mythri-garaga**
### Summary

This PR includes changes to publish service connect stats in `PublishMetricsAPI` payload that we currently send to the TACS.

### Implementation details

- `agent/tcs/client`:
     -  Updated `publishMetrics()` to publish Service Connect metrics every 60s. Currently task metrics are published at 20s interval therefore Service Connect metrics will included in Task metrics every 3rd publish request send to TACS if task is SC enabled.
     -  The message size for publish request is capped at 10 tasks or at 1 MB
     -  Added `serviceConnectMetricsToPublishMetricRequests` method  loops over all service connect metrics for a given task metric.  Each service connect metric is added to the publish metrics request being to the TACS as long as the request size is under 1 MB. If adding a service connect metric to request exceeds the 1MB limit , then the SC metric will be added to a new request with other Task metadata `ClusterArn` , `TaskArn`,  `TaskDefinitionFamily`, `TaskDefinitionVersion`

**TODO**:  `GetInstanceMetrics()` will be updated to retrieve and update TaskMetric with service connect stats in the upcoming PR.

Previous TACS Payload change PR - https://github.com/aws/amazon-ecs-agent/pull/3162

### Testing

`make test`

New tests cover the changes:  yes

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/stats/engine.go [hunk 1]
```diff
@@ -441,6 +440,8 @@ func (engine *DockerStatsEngine) GetInstanceMetrics() (*ecstcs.MetricsMetadata,
 			continue
 		}
 
+		// TODO [SC]: Call GetServiceConnectStats() and add it to taskMetrics with corresponding taskArn
```
Comment (YashdalfTheGray) [inline agent/stats/engine.go:443]: Question; and potentially because I don't have context, is this todo supposed to be here?

File: agent/tcs/client/client.go [hunk 2]
```diff
@@ -242,6 +263,44 @@ func (cs *clientServer) metricsToPublishMetricRequests() ([]*ecstcs.PublishMetri
 	return requests, nil
 }
 
+// serviceConnectMetricsToPublishMetricRequests loops over all the SC metrics in a
+// task metric to add SC metrics until the message size is within 1 MB.
+// If adding a SC metric to the message exceeds the 1 MB limit, it will be sent in the new message
+func (cs *clientServer) serviceConnectMetricsToPublishMetricRequests(requestMetadata *ecstcs.MetricsMetadata, taskMetric *ecstcs.TaskMetric,
+	messageTaskMetrics []*ecstcs.TaskMetric, requests []*ecstcs.PublishMetricsRequest) (*ecstcs.TaskMetric, []*ecstcs.TaskMetric, []*ecstcs.PublishMetricsRequest) {
+	tempTaskMetric := *taskMetric
```
Comment (angelcar) [inline agent/tcs/client/client.go:271]: why is this needed?

File: agent/tcs/client/client.go [hunk 3]
```diff
@@ -242,6 +263,44 @@ func (cs *clientServer) metricsToPublishMetricRequests() ([]*ecstcs.PublishMetri
...
+	tempTaskMetric.ServiceConnectMetricsWrapper = tempTaskMetric.ServiceConnectMetricsWrapper[:0]
+
+	for j, serviceConnectMetric := range taskMetric.ServiceConnectMetricsWrapper {
```
Comment (angelcar) [inline agent/tcs/client/client.go:274]: I don't think this loop is doing anything? you're emptying out `tempTaskMetric.ServiceConnectMetricsWrapper` in linve 277 above.

File: agent/tcs/client/client.go [hunk 4]
```diff
@@ -242,6 +263,44 @@ func (cs *clientServer) metricsToPublishMetricRequests() ([]*ecstcs.PublishMetri
...
+		tempTaskMetric.ServiceConnectMetricsWrapper = append(tempTaskMetric.ServiceConnectMetricsWrapper, serviceConnectMetric)
+		messageTaskMetrics = append(messageTaskMetrics, &tempTaskMetric)
+		tempMessage, _ := jsonutil.BuildJSON(ecstcs.NewPublishMetricsRequest(requestMetadata, copyTaskMetrics(messageTaskMetrics)))
```
Comment (angelcar) [inline agent/tcs/client/client.go:277]: Hmm, I worry about the CPU & memory consumption here since `BuildJSON` will create a lot of garbage and takes processing time. Let's not do premature optimization, but keep an eye on this. We will need to load test & profile this to see if it's an issue or not.  (maybe add a TODO for that here)

Alternatively, JSON has a somewhat predictable overhead. We could make calculations based on the size of the go structs and then add JSON overhead as a percentage.

File: agent/stats/engine.go (hunk 1 above)
Comment (mythri-garaga) [reply agent/stats/engine.go:443]: Thought GetInstanceMetrics() would be right place to gather service connect metrics since it initializes TaskMetric which contains the ServiceConnectMetricsWrapper.

```
type TaskMetric struct {
	_ struct{} `type:"structure"`

	ClusterArn *string `locationName:"clusterArn" type:"string"`

	ContainerMetrics []*ContainerMetric `locationName:"containerMetrics" type:"list"`

	ServiceConnectMetricsWrapper []*GeneralMetricsWrapper `locationName:"serviceConnectMetricsWrapper" type:"list"`

	...
}
```

Will implement this todo in next PR:)

File: agent/tcs/client/client.go [hunk 5]
```diff
@@ -39,19 +41,28 @@ const (
...
 	// tasksInHealthMessage is the maximum number of tasks that can be sent in a message to the backend
 	tasksInHealthMessage = 10
+	// defaultPublishServiceConnectTicker is every 3rd time service connect metrics will be sent to the backend
+	// Task metrics are published at 20s interval, thus task's service metrics will be published 60s.
+	defaultPublishServiceConnectTicker = 3
+)
+
+var (
+	// publishMetricRequestSizeLimit is the maximum number of bytes that can be sent in a message to the backend
+	publishMetricRequestSizeLimit = 1000000
```
[the remark this answers is not in the document]
Comment (mythri-garaga) [reply agent/tcs/client/client.go]: yes...updated, thank you!!

File: agent/tcs/client/client.go [hunk 6]
```diff
@@ -39,19 +41,28 @@ const (
...
-	ctx                      context.Context
-	cancel                   context.CancelFunc
-	disableResourceMetrics   bool
-	publishMetricsInterval   time.Duration
+	statsEngine                 stats.Engine
+	doctor                      *doctor.Doctor
+	publishTicker               *time.Ticker
+	publishHealthTicker         *time.Ticker
+	pullInstanceStatusTicker    *time.Ticker
+	publishServiceConnectTicker int32
```
[the remark this answers is not in the document]
Comment (mythri-garaga) [reply agent/tcs/client/client.go]: Done.

File: agent/tcs/client/client.go [hunk 7]
```diff
@@ -215,18 +232,22 @@ func (cs *clientServer) metricsToPublishMetricRequests() ([]*ecstcs.PublishMetri
 		return requests, nil
 	}
 	var messageTaskMetrics []*ecstcs.TaskMetric
+	var requestMetadata *ecstcs.MetricsMetadata
 	numTasks := len(taskMetrics)
 
 	for i, taskMetric := range taskMetrics {
+		requestMetadata = copyMetricsMetadata(metadata, false)
+
+		// TODO [SC]: Check if SC is enabled for the task
```
[the remark this answers is not in the document]
Comment (mythri-garaga) [reply agent/tcs/client/client.go:240]: Yes, will take care of this TODO in the next PR. Thank you:)

File: agent/tcs/client/client_test.go [hunk 8]
```diff
@@ -309,6 +379,51 @@ func TestPublishOnceNonIdleStatsEngine(t *testing.T) {
...
+			_, exists := taskArns[*taskMetric.TaskArn]
+			if exists {
+				if len(taskMetric.ContainerMetrics) != 0 {
+					t.Errorf("Expected Container metrics to be empty, got %d", len(taskMetric.ContainerMetrics))
+				}
+			}
+			taskArns[*taskMetric.TaskArn] = true
+		}
+	}
+	if len(requests) != expectedRequests {
```
[the remark this answers is not in the document]
Comment (mythri-garaga) [reply agent/tcs/client/client_test.go]: Updated

File: agent/tcs/client/client.go [hunk 9]
```diff
@@ -171,7 +182,13 @@ func (cs *clientServer) publishMetrics() {
 	for {
 		select {
 		case <-cs.publishTicker.C:
-			err := cs.publishMetricsOnce()
+			var includeServiceConnectStats bool
+			atomic.AddInt32(&cs.publishServiceConnectTicker, 1)
```
[the remark this answers is not in the document]
Comment (mythri-garaga) [reply agent/tcs/client/client.go]: You are right atomic operation is not really required here...hence updated it to simple increment logic.

File: agent/tcs/client/client_test.go [hunk 10]
```diff
@@ -309,6 +379,51 @@ func TestPublishOnceNonIdleStatsEngine(t *testing.T) {
...
+	requests, err := cs.metricsToPublishMetricRequests(true)
+	if err != nil {
+		t.Fatal("Error creating publishmetricrequests: ", err)
+	}
+
+	taskArns := make(map[string]bool)
+	for _, request := range requests {
+		for _, taskMetric := range request.TaskMetrics {
+			_, exists := taskArns[*taskMetric.TaskArn]
+			if exists {
```
[the remark this answers is not in the document]
Comment (mythri-garaga) [reply agent/tcs/client/client_test.go:404]: This was just to test the part of task metric being sent in next message has empty `ContainerMetrics` since we don't want agent to send it more than once to TACS.

I can add validation to check that  `ContainerMetrics` in the first part of `taskMetric` is not empty.

File: agent/tcs/client/client.go (hunk 4 above)
Comment (mythri-garaga) [reply agent/tcs/client/client.go:276]: Agreed, will add a TODO for this.

File: agent/tcs/client/client_test.go [hunk 11]
```diff
@@ -198,6 +200,74 @@ func newNonIdleStatsEngine(numTasks int) *nonIdleStatsEngine {
 	return &nonIdleStatsEngine{numTasks: numTasks}
 }
 
+type serviceConnectStatsEngine struct {
+	numTasks int
+}
+
+func (engine *serviceConnectStatsEngine) GetInstanceMetrics(includeServiceConnectStats bool) (*ecstcs.MetricsMetadata, []*ecstcs.TaskMetric, error) {
```
[the remark this answers is not in the document]
Comment (mythri-garaga) [reply agent/tcs/client/client_test.go:207]: Since `serviceConnectStatsEngine` implements the `Engine` interface, we can change the definition of `GetInstanceMetrics` method.

File: agent/tcs/client/client.go (hunk 2 above)
Comment (mythri-garaga) [reply agent/tcs/client/client.go:270]: Hmm do you mean why we need a temp variable?

File: agent/tcs/client/client.go (hunk 3 above)
Comment (mythri-garaga) [reply agent/tcs/client/client.go:273]: hmm I don't understand....I think we need this loop to incrementally add the service connect metrics to temp  variable to keep the request size in check(probably, I misunderstood your question here)

Comment (angelcar) [reply agent/tcs/client/client.go:273]: Oh, nvm. Just realized the loop is over `taskMetric.ServiceConnectMetricsWrapper`. I was reading it as if it was over `tempTaskMetric.ServiceConnectMetricsWrapper`

File: agent/tcs/client/client.go (hunk 2 above)
Comment (angelcar) [reply agent/tcs/client/client.go:270]: yeah, I was misreading this code. nvm,

**Comment by angelcar** [comment]
One last question here. What takes precedence, the 10 tasks, or the 1MB? 

For example, let's say there are 20 tasks running in an instance, each producing worth 1.5 MB of metrics data. How many requests are sent in that case? Would be good to have  a UT to validate that scenario.

**Comment by mythri-garaga** [comment]
@angelcar 
> One last question here. What takes precedence, the 10 tasks, or the 1MB?
 which ever hits first! Suppose we have 10 tasks, for 10th task's metric only it's first service connect metric fits the 1 MB limit, we would these in the first request and then send the 10th task metric with remaining service connect metrics in a new request.

> For example, let's say there are 20 tasks running in an instance, each producing worth 1.5 MB of metrics data. How many requests are sent in that case? Would be good to have a UT to validate that scenario.

In this case it would be 30 requests. sure, can add the this test case

**Review by angelcar — APPROVED** [review]

**Review by YashdalfTheGray — DISMISSED** [review]

**Review by angelcar — APPROVED** [review]

**Review by YashdalfTheGray — APPROVED** [review]
